src/charm.py

```diff
@@ -1,177 +1,130 @@
-                for e in Path("/proc/1/environ").read_text().split("\x00")
-                if "KUBERNETES_SERVICE" in e
-            )
-        )
         # Authenticate against the Kubernetes API using a mounted ServiceAccount token
         kubernetes.config.load_incluster_config()
         # Test the service account we've got for sufficient perms
-        auth_api = kubernetes.client.RbacAuthorizationV1Api(kubernetes.client.ApiClient())
+        api = kubernetes.client.CoreV1Api(kubernetes.client.ApiClient())
 
         try:
-            auth_api.list_cluster_role()
+            api.list_namespaced_service(namespace=self.namespace)
         except kubernetes.client.exceptions.ApiException as e:
             if e.status == 403:
                 # If we can't read a cluster role, we don't have enough permissions
                 self.unit.status = BlockedStatus("Run juju trust on this application to continue")
                 return False
             else:
                 raise e
-
-        self._authed = True
         return True
-    
+
+    @property
+    def _layer(self) -> dict:
+        """Returns a pebble layer for Portainer"""
+        return {
+            "services": {
+                "portainer": {
+                    "override": "replace",
+                    "command": "/portainer --tunnel-port 30776",
+                }
+            },
+        }
+
+    @property
+    def _service(self) -> dict:
+        """Return a Kubernetes service setup for Portainer"""
+        return {
+            "namespace": self.namespace,
+            "body": kubernetes.client.V1Service(
+                api_version="v1",
+                metadata=kubernetes.client.V1ObjectMeta(
+                    namespace=self.namespace,
+                    name=self.app.name,
+                    labels={
+                            "io.portainer.kubernetes.application.stack": self.app.name,
+                            "app.kubernetes.io/name": self.app.name,
+                            "app.kubernetes.io/instance": self.app.name,
+                            "app.kubernetes.io/version": "ce-latest-ee-2.4.0",
+                    },
+                ),
+                spec=kubernetes.client.V1ServiceSpec(
+                    type="NodePort",
+                    ports=[
+                        kubernetes.client.V1ServicePort(
+                            name="http",
+                            port=9000,
+                            target_port=9000,
+                            node_port=30776,
+                        ),
+                        kubernetes.client.V1ServicePort(
+                            name="edge",
+                            port=8000,
+                            target_port=8000,
+                            node_port=30777,
+                        ),
+                    ],
+                    selector={
+                        "app.kubernetes.io/name": self.app.name,
+                    },
```

Comment:
In your previous code, the selector was what prevented the `NodePort` behaving as expected